Correct logging in ph5_validate like other ph5 tools

damhuonglan opened the pull request:
### What does this PR do?
ph5_validate build logging in a way that different with other ph5 tools' which makes others when they are called in unittest keep calling ph5_validate.log for their logging file if they don't have their own. It doesn't affect other tools' functions but make unittest checks failed. Currently the check in PR #383 are failed because of this problem.

This PR correct the logging in ph5_validate to be the same as others

### Checklist
- [ ] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] All tests pass.
- [ ] Any new or changed features have are documented.
- [ ] Changes have been added to `CHANGELOG.txt` .
- [ ] First time contributors have added your name to `CONTRIBUTORS.txt` .

dsentinel commented on ph5/utilities/ph5validate.py:
```diff
@@ -53,31 +55,13 @@ class PH5ValidateException(Exception):
 
 
 class PH5Validate(object):
-    def __init__(self, ph5API_object, ph5path,
-                 level, outfile):
+    def __init__(self, ph5API_object, ph5path):
```
Does removing outfile, remove the capability to have output to a file instead of stdout?

dsentinel replied:
Changing the call signature here?
Is this class used anywhere else?

damhuonglan replied:
> Does removing outfile, remove the capability to have output to a file instead of stdout?

No, this outfile is to setup log file. I setting log file at line https://github.com/PIC-IRIS/PH5/blob/dca1c146e90848bcf1f3d415fe1440a6f741f769/ph5/utilities/ph5validate.py#L842
You can see that unittest still be able to check the content of ph5_validate.log as before changing logging configuration.

damhuonglan replied:
> Is this class used anywhere else?

This is a validating tool, it isn't used anywhere else except for its unittest.

dsentinel replied:
What??!!?!
This was only written to be tested?

dsentinel replied:
This is used to validate ph5 before being archived!!!!

damhuonglan replied:
> This was only written to be tested?

No I means this is call when this ph5_validate tool is used from command line. This is not call from other files in PH5.

dsentinel commented on ph5/utilities/ph5validate.py:
```diff
@@ -847,14 +831,32 @@ def get_args():
+    else:
+        # Set up logging
+        # Write log to file
+        ch = logging.FileHandler(os.path.join('.', outfile))
```
This won't let a user set outfile to an absolute path, which should be allowed.

damhuonglan replied:
Should I change this to:
`ch = logging.FileHandler(outfile)`

dsentinel replied:
yes, if no path I think it assumes ./

dsentinel commented on ph5/utilities/ph5validate.py:
```diff
@@ -847,14 +831,32 @@ def get_args():
-    if args.level != "ERROR" and args.level != "WARNING" and \
-            args.level != "INFO":
-        raise ValueError("Invalid logging level.")
+    outfile = 'ph5_validate.log' if args.outfile is None else args.outfile
```
this is already the default
add choices to `parser.add_argument(` above.

damhuonglan replied:
I will change this to:
`outfile = args.outfile`

dsentinel replied:
> this is already the default
> add choices to `parser.add_argument(` above.

dsentinel replied:
Looks like you missed this.

damhuonglan replied:
How do you think about adding help for option -o above?

```
    parser.add_argument(
        "-o", "--outfile", action="store",
        default="ph5_validate.log", type=str, metavar="outfile",
        help=("To record logging detail according to level stated in option "
              "-l. Default is ph5_validate.log"))
```

dsentinel replied:
```    parser.add_argument(
        "-l", "--level", action="store", default="WARNING",
        type=str, metavar="level", choices=('error', 'debug....
        help=("Level of logging detail. Choose from ERROR, WARNING, or INFO"))
```

damhuonglan replied:
Thank you for your help. I misunderstood your comment. I thought you talked about argument -o instead of argument -l. I fixed it according to your recommendation.

dsentinel commented on ph5/utilities/ph5validate.py:
```diff
@@ -847,14 +831,32 @@ def get_args():
+        # Set up logging
+        # Write log to file
+        ch = logging.FileHandler(os.path.join('.', outfile))
+        ch.setLevel(logging.INFO)
+        if args.verbose is True:
+            ch.setLevel(logging.DEBUG)
+        # Add formatter
+        formatter = logging.Formatter(LOGGING_FORMAT)
+        ch.setFormatter(formatter)
+        LOGGER.addHandler(ch)
+        level = args.level.upper()
+        if level == "ERROR":
+            LOGGER.setLevel(logging.ERROR)
+        elif level == "WARNING":
+            LOGGER.setLevel(logging.WARNING)
+        elif level == "INFO":
+            LOGGER.setLevel(logging.INFO)
+        else:
+            raise PH5ValidateException("Invalid Logging Level %s" % level)
```
This is messy.
And does verbose just set log level to debug?

dsentinel replied:
why `LOGGER.setLevel()` and `ch.setLevel()`?

damhuonglan replied:
Yes, I see this is messy too. I wasn't sure about this part. I think I will change this part to use ch.setLevel() and verbose is the first priority before level.

dsentinel commented on ph5/utilities/ph5validate.py:
```diff
@@ -847,14 +831,32 @@ def get_args():
+    if not os.path.exists(PH5) and not os.path.exists(PH5 + '.ph5'):
+        LOGGER.error("{0} not found.".format(PH5))
+        sys.exit()
+    else:
```
This shouldn't be an else. If above exits

damhuonglan replied:
I will remove the else

damhuonglan replied:
@MaevaP This part of code is outdated. I already removed the `else`.

dsentinel commented on ph5/utilities/ph5validate.py:
```diff
@@ -847,14 +831,33 @@ def get_args():
-    if args.level != "ERROR" and args.level != "WARNING" and \
-            args.level != "INFO":
-        raise ValueError("Invalid logging level.")
+    outfile = args.outfile
```
```suggestion

```

damhuonglan replied:
You was correct about args.outfile already has the default 'ph5_validate.log so I change line 834. But I don't understand why you want to remove the line 834 which defines  variable outfile. When I removed that as you suggested, I got the error "global name 'outfile' is not defined".

dsentinel commented on ph5/utilities/ph5validate.py:
```diff
@@ -863,17 +866,14 @@ def main():
         validation_blocks.extend(ph5validate.check_event_t())
         with open(args.outfile, "w") as log_file:
             for vb in validation_blocks:
-                vb.write_to_log(log_file,
-                                args.level)
+                vb.write_to_log(log_file, args.level.upper())
```
This should use the logger

damhuonglan replied:
[write_to_log() ](https://github.com/PIC-IRIS/PH5/blob/f4e8f1b4ebd79ef3c3c65de7d797e4062023cb26/ph5/utilities/ph5validate.py#L39)will write into ph5_validate.log in a special format for each station, whole event_t, response_t or experiment_t check with the header following with info/warning/error depending on the level selected.
Ex:

```
-=-=-=-=-=-=-=-=-
Experiment_t
1 error, 0 warning, 0 info
-=-=-=-=-=-=-=-=-
ERROR: Experiment_t does not exist. run experiment_t_gen to create table
-=-=-=-=-=-=-=-=-
Station 1111 Channel 1
2 error, 0 warning, 0 info
-=-=-=-=-=-=-=-=-
ERROR: Channel longitude 327519.7 not in range [-180,180]
ERROR: Channel latitude 3773952.7 not in range [-90,90]
-=-=-=-=-=-=-=-=-
Station 1111 Channel 1
2 error, 0 warning, 0 info
-=-=-=-=-=-=-=-=-
ERROR: Channel longitude 327519.7 not in range [-180,180]
ERROR: Channel latitude 3773952.7 not in range [-90,90]
```

I can remove the format for the file logger and still use write_to_log() function but using logger to write instead of opening the ph5_validate.log file to write into. However, the new way will write every info, warning, error to both screen and file; while with the old way I can log important info like which table are being checked or serious error to screen, user will read in the log file for all info, warning, error. DG is already used to the old way.   

Example of log info on screen:

```
[2020-08-12 16:40:50,103] - ph5.utilities.ph5validate - INFO: Validating Experiment_t
[2020-08-12 16:40:50,104] - ph5.utilities.ph5validate - INFO: Validating Array_t
[2020-08-12 16:40:50,418] - ph5.utilities.ph5validate - INFO: Validating Event_t
[2020-08-12 16:40:50,419] - ph5.utilities.ph5validate - WARNING: Event_t table not found. Did this experiment have shots???

Warnings, Errors and suggestions written to logfile: ph5_validate.log
```

If I change to use logger for everything, all will be printed to screen and make it harder to see important information. Please tell me if you want to proceed it that way.

dsentinel replied:
This isn't good, but leave it.

dsentinel commented:
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
add choices to parser.add_argument( above.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

damhuonglan commented on ph5/utilities/ph5validate.py:
```diff
@@ -846,15 +829,25 @@ def get_args():
+        sys.exit()
     if args.verbose is True:
         LOGGER.parent.handlers[0].setLevel(logging.DEBUG)
+    # Set up logging
```
I'm about to remove adding file handler to logger from line 839 - 849 here because the content of outfile is created by write_to_log(), not by logger. Is this ok for me to do that?

damhuonglan commented on ph5/utilities/ph5validate.py:
```diff
@@ -846,15 +829,25 @@ def get_args():
+        sys.exit()
     if args.verbose is True:
         LOGGER.parent.handlers[0].setLevel(logging.DEBUG)
+    # Set up logging
+    # Write log to file
+    ch = logging.FileHandler(outfile)
+    level = args.level
+    if level == "ERROR":
+        ch.setLevel(logging.ERROR)
+    elif level == "WARNING":
+        ch.setLevel(logging.WARNING)
+    elif level == "INFO":
+        ch.setLevel(logging.INFO)
+    LOGGER.addHandler(ch)
+
```
I'm about to remove adding file handler to logger from line 839 - 849 here because the content of outfile is created by write_to_log(), not by logger. Is this ok for me to do that?

MaevaP replied:
I think you should keep that section of code because that is where you create the handler and formatter for your log file and where you set the formatter and level of logging details.

damhuonglan replied:
I will keep this code.

MaevaP reviewed (changes requested):
Please see my suggested changes and comments.

damhuonglan commented on ph5/utilities/ph5validate.py:
```diff
@@ -847,14 +831,32 @@ def get_args():
-    if args.level != "ERROR" and args.level != "WARNING" and \
-            args.level != "INFO":
-        raise ValueError("Invalid logging level.")
+    outfile = 'ph5_validate.log' if args.outfile is None else args.outfile
```
This part of code is outdated.

damhuonglan commented:
@MaevaP Thanks for reviewing the PR for me. However some of your reviews are seem to be on the lines that no longer exist. I add  "This part of code is outdated" under those lines.
I changed outfile to args.outfile according to your recommendation, added help for outfile and kept the line for setting formatter and level for logging.

MaevaP reviewed (approved):
That looks good to me.
